src/renderer/Screencopy.cpp

```diff
@@ -0,0 +1,482 @@
+#include "Screencopy.hpp"
+#include "../helpers/Log.hpp"
+#include "../helpers/MiscFunctions.hpp"
+#include "../core/hyprlock.hpp"
+#include "../core/Egl.hpp"
+#include "../config/ConfigManager.hpp"
+#include "wlr-screencopy-unstable-v1.hpp"
+#include <EGL/eglext.h>
+#include <cstdint>
+#include <gbm.h>
+#include <unistd.h>
+#include <sys/mman.h>
+#include <cstring>
+#include <libdrm/drm_fourcc.h>
+#include <GLES3/gl32.h>
+#include <GLES3/gl3ext.h>
+#include <GLES2/gl2ext.h>
+
+static PFNGLEGLIMAGETARGETTEXTURE2DOESPROC glEGLImageTargetTexture2DOES = nullptr;
+static PFNEGLQUERYDMABUFMODIFIERSEXTPROC   eglQueryDmaBufModifiersEXT   = nullptr;
+
+//
+std::string CScreencopyFrame::getResourceId(COutput* output) {
+    return std::format("screencopy:{}-{}x{}", output->stringPort, output->size.x, output->size.y);
+}
+
+CScreencopyFrame::CScreencopyFrame(COutput* output) : m_output(output) {
+    m_resourceID = getResourceId(m_output);
+
+    captureOutput();
+
+    static auto* const PSCMODE = (Hyprlang::INT* const*)g_pConfigManager->getValuePtr("general:screencopy_mode");
+    if (**PSCMODE == 2 || **PSCMODE == 0)
+        m_frame = std::make_unique<CSCDMAFrame>(m_sc);
+    else
+        m_frame = std::make_unique<CSCSHMFrame>(m_sc);
+}
+
+void CScreencopyFrame::captureOutput() {
+    m_sc = makeShared<CCZwlrScreencopyFrameV1>(g_pHyprlock->getScreencopy()->sendCaptureOutput(false, m_output->output->resource()));
+
+    m_sc->setBufferDone([this](CCZwlrScreencopyFrameV1* r) {
+        Debug::log(TRACE, "[sc] wlrOnBufferDone for {}", (void*)this);
+        if (!m_frame) {
+            onFailed();
+            return;
+        }
+
+        if (!m_frame->onBufferDone() || !m_frame->m_wlBuffer) {
+            Debug::log(ERR, "onBufferDone failed");
+            onFailed();
+            return;
+        }
+
+        m_sc->sendCopy(m_frame->m_wlBuffer->resource());
+
+        Debug::log(TRACE, "[sc] wlr frame copied");
+    });
+
+    m_sc->setFailed([this](CCZwlrScreencopyFrameV1* r) {
+        Debug::log(ERR, "[sc] wlrOnFailed for {}", (void*)r);
+
+        m_frame.reset();
+
+        onFailed();
+    });
+
+    m_sc->setReady([this](CCZwlrScreencopyFrameV1* r, uint32_t, uint32_t, uint32_t) {
+        Debug::log(TRACE, "[sc] wlrOnReady for {}", (void*)this);
+
+        if (!m_frame || !m_frame->onBufferReady(m_asset)) {
+            Debug::log(ERR, "onBufferReady failed");
+            onFailed();
+            return;
+        }
+
+        m_sc.reset();
+    });
+}
+
+void CScreencopyFrame::onFailed() {
+    static auto* const PSCMODE = (Hyprlang::INT* const*)g_pConfigManager->getValuePtr("general:screencopy_mode");
```

Comment:
didnt we move to the new C++-style vars from hyprlang in hyprlock? If not, might be something for another MR.